ppoprf/src/ppoprf.rs

```diff
@@ -118,6 +118,14 @@ impl ServerPublicKey {
     let md = md_pk.decompress().unwrap();
     Ok(Point::from(b + md))
   }
+
+  pub fn serialize_to_bincode(&self) -> Result<Vec<u8>, bincode::Error> {
```

Comment:
Please add tests for these functions. Just a simple check for the expected results with good and bad examples. If you intend to record the output in a client, it's important to check that the format remains compatible.

Reply:
I would argue against this for the following reasons:

- Extensive test coverage is already included in serde/bincode crates
- Having raw byte data (the output of bincode serialization) is not very readable/easy to visually validate
- If the `ServerPublicKey` struct changes, then the output of the function changes too. If we wanted to update the test to reflect that, we would have to write an example function to execute this function, and update the test with the output.

Reply:
Right, I'm not particularly concerned about bincode itself failing to serialize the structure. I was worried something about the bincode config could change, and it wouldn't be obvious looking at our crate.

Having to update tests does make deliberately changing the encoding or the `ServerPublicKey` struct more expensive. I come down in favour because I hope we are nearly done with changes like that, so guarding against inadvertent breakage seems worthwhile. It being more work to change is also a reminder to document it as a breaking API change, and having tests already written makes it easier to debug serialization issues; there's a place to paste the serialization you expect to work.

I appreciate your pushing back through! Doing things for considered reasons instead of one person's say so generally improves software quality. :)

Reply:
I think your concerns are reasonable. If we either update to a version of bincode with breaking changes (sort of unlikely IMO, but still something we should consider; better to be safe than sorry), or change the format of our serializable structs, then having some fail-safes to prevent serialization related breakage would be beneficial. 

Although, I'm still not a big fan of having huge byte data arrays in our test suite. I found a nice snapshot testing crate called [insta](https://insta.rs/). It will handle the storage of initial test output via a snapshot, and compare future test invocations within the snapshot. Snapshot testing could be a good standard for any serializable structs in our projects. I updated my nested-sta-rs PR to include serialization tests for NestedMessage.

Reply:
Cute! I was hoping `cargo insta review` would diff the debug format of the structs, but no such luck :/

With this example I can confirm it's using the FixedintEncoding, so the format of the test slice is:
- `base_pk` length, 64-bit  little-endian
- `base_pk` value, 44 byte base64-encoded string
- `md_pks` container element count (8), 64-bit little endian
- 8 `mk_pks` encoded as:
  - md value, 8-bit
  - length, 64-bit little endian
  - value, 44-bytes base64-encoded string
 
For a total of 8 + 44 + 8 + 8*(1 + 8 + 44) = 484 bytes.

I suspect the base64 value encodings are some interaction with serde's derive macro. Perhaps reshuffling the imports would help with that. As it is, bincode isn't saving much over json serialisation.

Thanks for adding tests!

Reply:
It probably doesn't save much for the `ServerPublicKey`, but for `NestedMessage` it does, so perhaps it makes sense to use bincode for consistency. I would also imagine bincode parsing is slightly faster than JSON. thx for sharing your concerns. schema changes are one of the bigger challenges with bincode